### ISSUE TITLE
Snapping Nearest Circle While Drawing

### ISSUE THREAD

**Issue by seyhunak** [Tinghui/HUIPatternLockView-Swift#4]
@Tinghui there was a problem in drawing lines. Lines should be snap close circle, when drawing.
How is can be possible?

Thanks.

**Comment by Tinghui** [comment]
I don't understand what you mean "Lines should be snap close circle", do you have screenshots about the problem?

In HUIPatternLockView, while drawing the line, if the touch point is in the frame of a normal state dot, this dot will be highlighted and added into the pattern. And dot's center will became a part of the line. This is default feature to help highlight a dot more easily while user is drawing the pattern.

You can checkout the implementation of `updateLinePathWithPoint` function for this feature. 

Or your can checkout the demo project, it has sample codes for custom drawing. Hope it’s helpful.

**Comment by seyhunak** [comment]
Hello @Tinghui 

I've added a animation to explain my case. 
While I started to draw from [0] to [3] skipped so then [6] pointed finally.
I need to snap circle [3] too, while going to [6] 

So, it creates better user experience to draw.
Can you help me about this issue?
Thanks 

Please see attached:

![animation](https://cloud.githubusercontent.com/assets/170820/18302618/7e3cc848-74e1-11e6-8102-45c39910fd57.gif)

**Comment by Tinghui** [comment]
I think I know what's you need. 

But it seems this is an unusual feature. Because when user is drawing from [0] to [6], if he's not specifically draw on the dot [3], this most possibly means user don't want the dot [3]. If we snap dot [3] automatically, it's not the user wants. 

So I'm not going to implement this feature in the repository, but if you really need this feature, you can fork this repository, and modify the implementation of `updateLinePathWithPoint` function to achieve it.

Basically, in `updateLinePathWithPoint`, you need do these:
1. make a line segment with the last two points in `linePath`
2. check if any dot in `normalDots` is crossed by the line segment.
3. if has one, insert dot's center at the index `linePathPointsCount - 1` of `linePath`
4. set `dot.highlighted = true`
5. add dot into highlightedDots `highlightedDots.append(dot)`
6. remove dot from normalDots `if let index = normalDots.index(of: dot) { normalDots.remove(at: index) }`

**Comment by Tinghui** [comment]
@seyhunak The code is here, you can check whether it's what you want.

The code is in Swift 3.0, if you are using swift 2.3, `fileprivate` should be changed to `private`  

``` swift
    // Detect whether two line segments intersected
    // From: smac89's answer in http://stackoverflow.com/questions/563198/how-do-you-detect-where-two-line-segments-intersect
    private func isIntersectedSegments(p0: CGPoint, p1: CGPoint, p2: CGPoint, p3: CGPoint) -> Bool {
        let s1_x = p1.x - p0.x
        let s1_y = p1.y - p0.y
        let s2_x = p3.x - p2.x
        let s2_y = p3.y - p2.y
        let s = (-s1_y * (p0.x - p2.x) + s1_x * (p0.y - p2.y)) / (-s2_x * s1_y + s1_x * s2_y)
        let t = ( s2_x * (p0.y - p2.y) - s2_y * (p0.x - p2.x)) / (-s2_x * s1_y + s1_x * s2_y)
        return s >= 0.0 && s <= 1.0 && t >= 0.0 && t <= 1.0
    }

    private func normalDotCrossedBySegment(start: CGPoint, end: CGPoint) -> Dot? {
        for dot in normalDots {
            let leftTop = CGPoint(x: dot.frame.minX, y: dot.frame.minY)
            let rightTop = CGPoint(x: dot.frame.maxX, y: dot.frame.minY)
            let leftBottom = CGPoint(x: dot.frame.minX, y: dot.frame.maxY)
            let rightBottom = CGPoint(x: dot.frame.maxX, y: dot.frame.maxY)
            if isIntersectedSegments(p0: start, p1: end, p2: leftTop, p3: rightTop)
                || isIntersectedSegments(p0: start, p1: end, p2: leftTop, p3: leftBottom)
                || isIntersectedSegments(p0: start, p1: end, p2: leftBottom, p3: rightBottom)
                || isIntersectedSegments(p0: start, p1: end, p2: leftTop, p3: rightBottom) {
                return dot
            }
        }

        return nil
    }

    fileprivate func updateLinePathWithPoint(_ point: CGPoint) -> Void {
        let linePathPointsCount = linePath.count
        let segmentStart: CGPoint? = (linePathPointsCount >= 2 ? linePath[linePathPointsCount - 2] : nil)
        let segmentEnd: CGPoint? = (linePathPointsCount >= 2 ? linePath[linePathPointsCount - 1] : nil)

        if var dot = normalDotContainsPoint(point) {
            if (linePathPointsCount <= 0) {
                //if no any points in linePath. use this dot's center to be the linePath start and end point
                linePath.append(dot.center)
                linePath.append(dot.center)
            }
            else {
                //else insert a new point into the path
                linePath.insert(dot.center, at: linePathPointsCount-1)
            }

            //mark this dot as highlighted
            dot.highlighted = true
            highlightedDots.append(dot)
            if let index = normalDots.index(of: dot) {
                normalDots.remove(at: index)
            }
        }
        else if let start = segmentStart
            , let end = segmentEnd
            , var dot = normalDotCrossedBySegment(start: start, end: end) {
            linePath.insert(dot.center, at: linePathPointsCount-1)
            //mark this dot as highlighted
            dot.highlighted = true
            highlightedDots.append(dot)
            if let index = normalDots.index(of: dot) {
                normalDots.remove(at: index)
            }
        }
        else {
            if (linePathPointsCount == 0) {
                //linePath must start with a dot's center
                return
            }
            else if (linePathPointsCount == 1) {
                //if linePath has a start point, this point is treat as end point
                linePath.append(point)
            }
            else {
                //if line path has at least two points. always use this point to update the end point
                linePath[linePathPointsCount-1] = point
            }
        }
    }
```

**Comment by seyhunak** [comment]
Thanks, using Swift 2.3. Working great!
